src/bio/bio_context.c

```diff
@@ -164,6 +164,98 @@ blob_msg_close(void *msg_arg)
 	spdk_blob_close(arg->bma_blob, blob_cb, arg->bma_cp_arg);
 }
 
+static int
+bio_check_bs(struct bio_blobstore *bbs)
+{
+	int rc = 0;
+
+	D_ASSERT(bbs != NULL);
+	ABT_mutex_lock(bbs->bb_mutex);
+	if (bbs->bb_bs == NULL) {
+		D_ERROR("Blobstore %p is closed, fail request.\n", bbs);
+		rc = -DER_NO_HDL;
+		goto out;
+	}
+
+	if (bbs->bb_state == BIO_BS_STATE_TEARDOWN ||
+	    bbs->bb_state == BIO_BS_STATE_OUT) {
+		D_ERROR("Blobstore %p is teardown, reject request.\n", bbs);
+		rc = -DER_STALE;
```

Comment:
Agreed, seems DER_DOS is better.

Yes, it's ok to release the mutex. Once the bs state is in TEARDOWN or OUT, there won't be any requests from upper layer. Even if there is any request for whatever reason, we should just reject it and return failure, because it doesn't make any sense to operate on a TEARDOWN/OUT blobsotre.

The bs state is allowed to be changed by others once the mutex is release here: If the state is changed from "TEARDOWN/OUT" to other states, there is nothing wrong because the request is already being rejected; If the state is changed from "FAULTY" to "TEARDOWN", the in-flight request holds a 'bs_holdings' to prevent the blobstore being torn down.